File: AtCoder/ABC-D/056probD.cpp
#include <bits/stdc++.h>
typedef long long ll;
#define REP(i, n) for(int i = 0; i < (int)(n); i++)
#define FOR(i, a, b) for(int i = (int)(a); i < (int)(b); i++)
using namespace std;

int main() {
    int N, K; cin >> N >> K;
    int A[N]; REP(i, N) cin >> A[i];

    int dp1[N][K];
    REP(i, K) dp1[0][i] = 0;
    dp1[0][0] = 1;
    REP(n, N-1) REP(k, K){
        if (k < A[n]){
            dp1[n+1][k] = dp1[n][k];
        }
        else{
            dp1[n+1][k] = dp1[n][k] | dp1[n][k-A[n]];
        }
    }

    int dp2[N][K];
    REP(i, K) dp2[N-1][i] = 0;
    dp2[N-1][0] = 1;
    for (int n=N-1; n>0; n--) REP(k, K){
        if (k < A[n]){
            dp2[n-1][k] = dp2[n][k];
        }
        else{
            dp2[n-1][k] = dp2[n][k] | dp2[n][k-A[n]];
        }
    }

    int ans = 0;
    REP(n, N){
        if (K-A[n] <= 0) continue;
        bool need = false;
        int k2 = K-1;
        REP(k1, K){
            if (!dp1[n][k1]) continue;
            while (!dp2[n][k2] || k1+k2>=K){
                k2 -= 1;
            }
            if (k1+k2 >= K-A[n]){
                need = true;
                break;
            }
        }
        if (!need) ans += 1;
    }
    cout << ans << endl;
    
}

File: AtCoder/ABC-D/056probD.py
import sys
input = sys.stdin.readline


def main():
    N, K = map(int, input().split())
    A = list(map(int, input().split()))

    dp1 = [set() for _ in range(N)]
    dp1[0].add(0)
    for n in range(N-1):
        for pm in dp1[n]:
            dp1[n+1].add(pm)
            if pm + A[n] < K:
                dp1[n+1].add(pm+A[n])

    dp2 = [set() for _ in range(N)]
    dp2[N-1].add(0)
    for n in reversed(range(1, N)):
        for pm in dp2[n]:
            dp2[n-1].add(pm)
            if pm+A[n] < K:
                dp2[n-1].add(pm+A[n])

    ans = 0
    for n, a in enumerate(A):
        b = K - a
        if b <= 0: continue
        need = False
        B1 = sorted(dp1[n])
        B2 = sorted(dp2[n], reverse=True)
        i2 = 0
        for b1 in B1:
            while B2[i2] + b1 >= K:
                i2 += 1
                if i2 == len(B2): break
            if i2 == len(B2): break
            if b1+B2[i2] < b:
                continue
            else:
                need = True
                break
        if not need: ans += 1
    print(ans)


if __name__ == "__main__":
    main()